iroh-dns-server/src/http.rs

```diff
@@ -0,0 +1,191 @@
+        .route("/dns-query", get(doh::get).post(doh::post))
+        .route(
+            "/pkarr/:key",
+            get(pkarr::get).put(pkarr::put.layer(rate_limit)),
+        )
+        .route("/healthcheck", get(|| async { "OK" }))
+        .route("/", get(|| async { "Hi!" }))
+        .with_state(state);
+
+    // configure app
+    let app = router
+        .layer(cors)
+        .layer(trace)
+        .route_layer(middleware::from_fn(metrics_middleware));
+
+    let mut tasks = JoinSet::new();
+
+    // launch http
+    if let Some(config) = http_config {
+        let app = app.clone();
+        let addr = SocketAddr::from((Ipv4Addr::UNSPECIFIED, config.port));
+        info!("HTTP server listening on {addr}");
+        let fut =
+            axum_server::bind(addr).serve(app.into_make_service_with_connect_info::<SocketAddr>());
+        tasks.spawn(fut);
+    };
+
+    // launch https
+    if let Some(config) = https_config {
+        let addr = SocketAddr::from((Ipv4Addr::UNSPECIFIED, config.port));
+        info!("HTTPS server listening on {addr}");
+        let acceptor = {
+            let cache_path = Config::data_dir()?
+                .join("cert_cache")
+                .join(config.cert_mode.to_string());
+            tokio::fs::create_dir_all(&cache_path)
+                .await
+                .with_context(|| format!("failed to create cert cache dir at {cache_path:?}"))?;
+            config
+                .cert_mode
+                .build(
+                    &config.domain,
+                    cache_path,
+                    config.letsencrypt_contact,
+                    config.letsencrypt_prod,
+                )
+                .await?
+        };
+        let fut = axum_server::bind(addr)
+            .acceptor(acceptor)
+            .serve(app.into_make_service_with_connect_info::<SocketAddr>());
+        tasks.spawn(fut);
+    }
+
+    loop {
+        tokio::select! {
+            // todo: graceful cancellation
+            _ = cancel.cancelled() => tasks.abort_all(),
+            res = tasks.join_next() => match res {
+                None => break,
+                Some(res) => res??,
+            }
+        }
+    }
+
+    Ok(())
+}
+
+async fn metrics_middleware(req: Request, next: Next) -> impl IntoResponse {
+    let start = Instant::now();
+    // let path = if let Some(matched_path) = req.extensions().get::<MatchedPath>() {
+    //     matched_path.as_str().to_owned()
+    // } else {
+    //     req.uri().path().to_owned()
+    // };
+    // let method = req.method().clone();
+    //
+    let response = next.run(req).await;
+    //
+    let latency = start.elapsed().as_millis();
+    let status = response.status();
+    inc_by!(Metrics, http_requests_duration_ms, latency as u64);
+    inc!(Metrics, http_requests);
+    if status.is_success() {
+        inc!(Metrics, http_requests_success);
+    } else {
+        inc!(Metrics, http_requests_error);
+    }
+    //
+    // let labels = [
```

Comment:
?

Reply:
this is copypasted from the axum-telemetry example. opentelemetry and prometheus can associate counters and histograms to labels, which allows to e.g. see how many PUT vs GET requests occured, or the request durations for a specific path, etc. `iroh-metrics` does not support this at the moment. cc @Arqu - maybe you know more what we are missing to support this? Not a high priority atm I think, but would be nice to have also for metrics in iroh (e.g. attach document id to doc metrics to thus track transfer etc per doc in addition to global),